Fix UI navigation breaking on invisible controls

Rindbee opened the pull request:
Use `is_visible_in_tree()` instead of `is_visible()` to check for invisible controls.

Fix #95747.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

Rindbee committed:
Fix ui navigation break

Use `is_visible_in_tree()` instead of `is_visible()` to check for
invisible controls.

KoBeWi commented:
I see some checks for `is_set_as_top_level()` too. Seems inconsistent.

KoBeWi reviewed (approved):
Fixes the issue, the `is_set_as_top_level()` can be looked at later.

Rindbee commented:
The previous proxy logic felt a bit convoluted, so let's make it simpler.

`is_set_as_top_level()` is consistent, because as long as the control is in the canvas, `data.RI` is equivalent to `is_set_as_top_level() || !data.parent_control`, but the latter is more reliable, especially when entering and exiting the canvas is possible.

Repiteo commented:
Needs a rebase

Rindbee commented:
Sorry, `data.RI` is equivalent to `is_set_as_top_level() || no ancestor control`. The incorrect replacement has been reverted.

akien-mga commented:
Such changes to core Control logic are pretty risky, so I'm not comfortable merging it now during 4.4 beta phase. We can aim to merge it early on for the 4.5 dev cycle.

It would be good to have unit tests added, ideally first confirming that there's a bug, and then validating this fixes them without introducing regressions.

Rindbee committed:
Add unit tests for control focus navigation

Users can navigate focus between controls using `ui_focus_next` or
`ui_focus_prev actions`. Add unit tests for these behaviors.

Repiteo commented:
Thanks!
